lego/components.py

```diff
@@ -0,0 +1,144 @@
+"""
+Component object provides the API which tests and libs will use to run code on the component.
+"""
+from __future__ import annotations
+from typing import Optional, Type, TypeVar
+from types import TracebackType
+import abc
+import socket
+import ipaddress
+
+import rpyc
+
+from .connections import BaseConnection, RPyCConnection
+
+T = TypeVar('T', bound='BaseComponent')  # pylint: disable=invalid-name
+
+
+class BaseComponent(metaclass=abc.ABCMeta):
+    """
+    BaseComponent class is the base class of all other components,
+    In order to extend the API implement a wrapping component, and to provide more
+    complex functionality add appropriate Lib.
+    """
+
+    def __init__(self, connection: BaseConnection) -> None:
```

Comment:
The component doesn't receive a connection but creates it.
The component receives the hostname and in some cases other values like user/password/port.
This method is abstract
```suggestion
    @abc.abstractmethod
    def __init__(self, hostname) -> None:
```

Reply:
I had in mind that every component will start in it's __init__ function the appropriate connection and then pass it to BaseComponent with a call to super().__init__(connection).
Therefore, in BaseComponent.__exit__ we could close the received connection.

If we are not passing connection, and we don't have any logic in BaseComponent.__init__ why should we even have it?.